Feat/weak attribute

susundberg opened the pull request:
Hi!

Here is pull request for issue #43. I have tested it only with gcc (i guess clang will be tested with the CI) - and used in my own project.

ekeroc commented:
Could I ask you why does need the __attribute__((weak)) symbol ?

susundberg commented:
Hi! I tried to reason the need in [issue 43](https://github.com/meekrosoft/fff/issues/43) . 

In short, i have a set of files (generated), that define fake functions for all real functions. This is compiled to be library -- say "libfake.a". Now when linking a test binary, i have maybe one or two real functions, and this binary is linked against the fakelib (i have many of those binaries - one for each set of real functions!). Now without the __weak__ one gets linker error -- the real function is defined also in fakelib. With __weak__ linker is instructed to select the other one with no error.

I could surely pick the files for compiling, but that would require additional "makefile"-logic -- now i can compile the whole fakelib once, and let the linker select suitable implementation for required functions.

meekrosoft commented:
I am a bit reluctant to merge this one...
I completely see the need - but not every compiler supports weak attributes using this syntax so it would break some users.  Is there another way?  Perhaps a preprocessor switch?  I imagining adding a helper macro to output something like this:

````

#IFDEF GCC_WEAK_FUNCTIONS
   #DEFINE WEAK_OPTION __attribute__((weak))
#ELSE
    #DEFINE WEAK_OPTION 
#

"#{return_type} WEAK_OPTION FUNCNAME(#{arg_val_list(arg_count)}#{varargs})"
````

wulfgarpro commented:
@susundberg, please resolve your conflicts before review moves forward.

susundberg commented:
Hi! Sorry for long delay.

I (force) pushed new update; i replaced the hardcoded __attribute__(weak) with #ifdef FFF_FUNCTION_ATTRIBUTE as i though it would fit better.

I updated also the generated fff.h but there seems to be whitespace differences also on unmodified master. I am running ubuntu 18.04 with 
```
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux-gnu]
```
Any suggestions what version to use (and maybe add that to readme?).

wulfgarpro commented:
@susundberg, the only whitespace changes I can see in your PR are good ones.  It looks like you've used your editor to strip them.

susundberg commented:
Hi! Sorry for the delay again. I actually did checked out the tests before making the pull request but decided not to go there since it was not clear me how they are structured etc.

I now went back to look how it goes - and would an examples/weak_linking/ with some tests included be sufficient? 

Happy holidays!

wulfgarpro commented:
Just define an example weak function and fake it, as was the need for this
functionality.

On 21 Dec. 2018 1:01 am, "Pauli Salmenrinne" <notifications@github.com>
wrote:

Hi! Sorry for the delay again. I actually did checked out the tests before
making the pull request but decided not to go there since it was not clear
me how they are structured etc.

I now went back to look how it goes - and would an examples/weak_linking/
with some tests included be sufficient?

Happy holidays!

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/meekrosoft/fff/pull/47#issuecomment-449008466>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AA6TtLl_9eB5jkvCxAuuW-DBaX14NIPNks5u65gtgaJpZM4TBg_E>
.

susundberg commented:
Uh sorry, ignore, force pushed after pulling to latest, examples still missing ..

susundberg commented:
Hi all again. I took a while but i added an example "weak_linking". Its compiled and runned on the normal command. The example itself is silly - as you can guess - but the Makefile includes three example cases where weak linking is required.

I did not make any (automatic) check that the compiling fails if the weak linking flag is not set. I did try manually though.

How does it look?
